---
title: "Ecoli-Human Spike-in Study (PRIDE ID PXD003881)"
author:
- name: Lieven Clement, Alexandre Segers
output:
  BiocStyle::html_document
---


# Background
We here use a publicly available spike-in study 
(PRIDE identifier: PXD003881 Shen et al. [2018]). E. Coli lysates were spiked
at five different concentrations (3%, 4.5%, 6%, 7.5% and 9% wt/wt)
in a stable human background (four replicates per treatment). The samples
were run on an Orbitrap Fusion mass spectrometer. Raw data files were processed
with MaxQuant (version 1.6.1.0, Cox and Mann [2008]) using default search
settings unless otherwise noted. Spectra were searched against the
UniProtKB/SwissProt human and E. Coli reference proteome databases (07/06/2018),
concatenated with the default Maxquant contaminant database. C
arbamidomethylation of Cystein was set as a fixed modification,
and oxidation of Methionine and acetylation of the protein amino-terminus
were allowed as variable modifications. In silico cleavage was set to use trypsin/P,
allowing two miscleavages. Match between runs was also enabled using default settings.
The resulting peptide-to-spectrum matches (PSMs) were filtered by MaxQuant at 1% FDR.

We first set the concertations for the different spike-ins.

```{r}
concentrations <- seq.int(1:5) * 1.5
names(concentrations) <- letters[1:5]
```

# Data

```{r}
library(tidyverse)
library(limma)
library(QFeatures)
library(msqrob2)
library(BiocParallel)
library(here)
```



```{r, warning=FALSE, message=FALSE}

pe <- readRDS(file = here("Data/Ecoli_Human/Ecoli_QFeatures.RDS"))
pe
```

We can extract the spikein condition from the raw file name.

```{r}
cond <- which(strsplit(colnames(pe)[[1]][1], split = "")[[1]] == "a") # find where condition is stored
colData(pe)$condition <- substr(colnames(pe), cond, cond) %>% unlist %>%  as.factor
```

We calculate how many non zero intensities we have per peptide.
This will be useful for filtering.

```{r}
rowData(pe[["peptideRaw"]])$nNonZero <- rowSums(assay(pe[["peptideRaw"]]) > 0)
```


Peptides with zero intensities are missing peptides and
should be represent with a `NA` value instead of `0`.
```{r}
pe <- zeroIsNA(pe,"peptideRaw")
```

## Information on species

In the spik-in study there are peptides from e.coli and human proteins.
The ecoli peptides are added at different concerntrations.

```{r}
ecoli <- read.delim(file = here("Data/Ecoli_Human/ecoli.fasta"))
human <- read.delim(file = here("Data/Ecoli_Human/human.fasta"))
```

```{r}
id <- list(ecoli = 'ecoli.fasta',
          human = 'human.fasta') %>%
  purrr::map(~{read_lines(.x) %>%
          {.[str_detect(.,'^>')]} %>%
          str_extract(.,'(?<=\\|).*(?=\\|)')})
```
## Data exploration

We can inspect the missingness in our data with the `plotNA()` function
provided with `MSnbase`. `r format(mean(is.na(assay(pe[["peptideRaw"]])))*100,digits=2)`%
of all peptide intensities are missing and for some peptides we de not
even measure a signal in any sample. The missingness is similar across samples.


```{r, cache= TRUE, eval=FALSE}
MSnbase::plotNA(assay(pe))  +
  xlab("Peptide index (ordered by data completeness)")
```


# Preprocessing

We normalize the data using `vsn` normalisation.
Note, that the data should not be log-transformed.


## Filtering

### Handling overlapping protein groups
In our approach a peptide can map to multiple proteins, as long as there is
none of these proteins present in a smaller subgroup.

```{r}
pe[["peptideRaw"]] <- pe[["peptideRaw"]][rowData(pe[["peptideRaw"]])$Proteins %in%
                                           smallestUniqueGroups(rowData(pe[["peptideRaw"]])$Proteins), ]
```

### Remove reverse sequences (decoys) and contaminants

We now remove the contaminants, peptides that map to decoy sequences
and proteins, which were only identified by peptides with modifications.

```{r}
pe[["peptideRaw"]] <- pe[["peptideRaw"]][rowData(pe[["peptideRaw"]])$Reverse != "+", ]
pe[["peptideRaw"]] <- pe[["peptideRaw"]][rowData(pe[["peptideRaw"]])$
Potential.contaminant != "+", ]
```

### Drop peptides that were only identified in one sample

We want to keep peptide that were observed at least twice.

```{r}
pe[["peptideRaw"]] <- pe[["peptideRaw"]][rowData(pe[["peptideRaw"]])$nNonZero >= 2, ]
nrow(pe[["peptideRaw"]])
```

We keep `r nrow(pe[["peptideRaw"]])` peptides upon filtering.


## Normalize the data using the vsn method
```{r}
pe <- logTransform(pe, base = 2, i = "peptideRaw", name = "peptideLog")
pe <- normalize(pe, 
                i = "peptideLog", 
                name = "peptideNorm", 
                method = "center.median")

```

```{r}
pe
```

## Full dataset

```{r}
set.seed(100)
sce <- pe[["peptideNorm"]]
names(assays(sce)) <- "logintensities"
```


```{r}
saveRDS(object = sce, file = here("Data/Ecoli_Human/Ecoli_Human_sce.RDS"))
write.csv(x = assay(sce,"logintensities"), file = here("Data/Ecoli_Human/Ecoli_Human_sce.csv"))
```


## Imputation of NAs

```{r}
source(here("R-functions", "Impute_MAR_MNAR.R"))
```

```{r}
set.seed(100)

for(i in c(1:10)){
  MISTOT10_MNAR25 <- imputation_NAs(data = assay(sce, "logintensities"), MISTOT = 0.1, MNAR = 0.25)
  MISTOT10_MNAR50 <- imputation_NAs(data = assay(sce, "logintensities"), MISTOT = 0.1, MNAR = 0.50)
  MISTOT10_MNAR75 <- imputation_NAs(data = assay(sce, "logintensities"), MISTOT = 0.1, MNAR = 0.75)

  
  saveRDS(MISTOT10_MNAR25, paste0(here("Data","Ecoli_Human"), "/Ecoli_Human_MISTOT10_MNAR25_",i,".RDS"))
  saveRDS(MISTOT10_MNAR50, paste0(here("Data","Ecoli_Human"), "/Ecoli_Human_MISTOT10_MNAR50_",i,".RDS"))
  saveRDS(MISTOT10_MNAR75, paste0(here("Data","Ecoli_Human"), "/Ecoli_Human_MISTOT10_MNAR75_",i,".RDS"))

  write.csv(x = MISTOT10_MNAR25$data[rowMeans(is.na(MISTOT10_MNAR25$data)) < 1,], 
            file = paste0(here("Data","Ecoli_Human"), "/Ecoli_Human_MISTOT10_MNAR25_",i,".csv"))
  write.csv(x = MISTOT10_MNAR50$data[rowMeans(is.na(MISTOT10_MNAR50$data)) < 1,], 
            file = paste0(here("Data","Ecoli_Human"), "/Ecoli_Human_MISTOT10_MNAR50_",i,".csv"))
  write.csv(x = MISTOT10_MNAR75$data[rowMeans(is.na(MISTOT10_MNAR75$data)) < 1,], 
            file = paste0(here("Data","Ecoli_Human"), "/Ecoli_Human_MISTOT10_MNAR75_",i,".csv"))
}
```



